### DISCUSSION TITLE
VRL move root to new field

### DISCUSSION THREAD

**Discussion by alexeynl** [vectordotdev/vrl#1057] [Q&A]
Is there any good way to move root (.) of the event to new field?

I tried the following

.newfield = del(.)

but next following assingment

.field1.subfield1 = "value"

get me the error

"querying a field of a non-object type is unsupported"

I understand that i can copy root to new field and next remove all others, but this looks not efficient way to do such simple thing...

**Comment by pront** [comment]
👋 Hi @alexeynl, see this [example](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiIyBSZW1vdmUgc29tZSBmaWVsZHNcbi54ID0gZGVsKC4pXG4ueSA9IC54LmZvbyIsImV2ZW50Ijp7Im1lc3NhZ2UiOiJIZWxsbyBWUkwiLCJmb28iOiJkZWxldGUgbWUiLCJodHRwX3N0YXR1cyI6IjIwMCJ9LCJpc19qc29ubCI6ZmFsc2UsImVycm9yIjpudWxsfQ%3D%3D). It assigns the whole top level object to a new path.

**Reply by alexeynl** [reply]
Hi @pront 
Just for clarification: I do not need just assing, i need MOVE content of event root to new field. This means that root can contains a lof fields before move and must become empty except new field after.
Furthermore if you will add 
```
.a.b = "value"
```
to your example you will encounter with the same 
"querying a field of a non-object type is unsupported"'
This is not accepted behavior, because in this case (after `del(.)` used) i need always to add garbage to my VRL code like this:
```
.a.b = {}
.a.b.c = {}
.a.b.c.d = "value"
```
instead of (in case i've not done  `del(.)` before)
```
.a.b.c.d = "value"
```

**Reply by pront** [reply]
Hey, from your description this sounds like VRL is working as designed. After the assignment `.a.b = {}` the compiler knows it is dealing with an object type. Also, worth taking a look at https://vector.dev/docs/reference/vrl/functions/#set. If you shared a playground link maybe it would become more obvious.

**Reply by alexeynl** [reply]
> If you shared a playground link maybe it would become more obvious.

**Example 1.** 
```
.field.subfield = "value1"
```
**VRL Playground**
https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiLmZpZWxkLnN1YmZpZWxkID0gXCJ2YWx1ZTFcIiIsImV2ZW50Ijp7Im1lc3NhZ2UiOiJIZWxsbyBWUkwiLCJmb28iOiJkZWxldGUgbWUiLCJodHRwX3N0YXR1cyI6IjIwMCJ9LCJpc19qc29ubCI6ZmFsc2UsImVycm9yIjpudWxsfQ%3D%3D
**Result:** No errors


**Example 2.** : Same thing but used del(.) first:
```
.newfield = del(.)
.field.subfield = "value1"
```
https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiLm5ld2ZpZWxkID0gZGVsKC4pXG4uZmllbGQuc3ViZmllbGQgPSBcInZhbHVlMVwiIiwiZXZlbnQiOnsibWVzc2FnZSI6IkhlbGxvIFZSTCIsImZvbyI6ImRlbGV0ZSBtZSIsImh0dHBfc3RhdHVzIjoiMjAwIn0sImlzX2pzb25sIjpmYWxzZSwiZXJyb3IiOm51bGx9
**Result:**
Getting error: `querying a field of a non-object type is unsupported`

**Reply by pront** [reply]
Added debug statements that should help clear up the confusion around how types change -> [here](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoidDEgPSB0eXBlX2RlZiguKVxuLm5ld2ZpZWxkID0gZGVsKC4pXG4uc2Vjb25kX3Jvb3RfdHlwZSA9IHR5cGVfZGVmKC4pXG4uZmlyc3Rfcm9vdF90eXBlID0gdDFcblxuIyB1bmRlZmluZWQsIGNhbm5vdCB0cmVhdCBhcyBvYmplY3Rcbi5maXJzdF9maWVsZF90eXBlID0gdHlwZV9kZWYoLmZpZWxkKVxuXG4jIGNvbnZlciB0byBvYmplY3Rcbi5maWVsZCA9IHt9XG4uc2Vjb25kX2ZpZWxkX3R5cGUgPSB0eXBlX2RlZiguZmllbGQpXG5cbiMgbm93IGNhbiB1c2UgcGF0aCBzeW50YXhcbi5maWVsZC5zdWJmaWVsZCA9IFwidmFsdWUxXCIiLCJldmVudCI6eyJtZXNzYWdlIjoiSGVsbG8gVlJMIiwiZm9vIjoiZGVsZXRlIG1lIiwiaHR0cF9zdGF0dXMiOiIyMDAifSwiaXNfanNvbmwiOmZhbHNlLCJlcnJvciI6bnVsbH0%3D).

**Reply by alexeynl** [reply]
> now can use path syntax

```
# now can use path syntax
.field.subfield = "value1"
```

No, you can't! This is main problem in this case! Try to replace:

```
.field.subfield = "value1" ->  .field.subfield.subsubfield = "value1"
```
in you example and you will get the same error `querying a field of a non-object type is unsupported`!

[VRL Playground](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoidDEgPSB0eXBlX2RlZiguKVxuLm5ld2ZpZWxkID0gZGVsKC4pXG4uc2Vjb25kX3Jvb3RfdHlwZSA9IHR5cGVfZGVmKC4pXG4uZmlyc3Rfcm9vdF90eXBlID0gdDFcblxuIyB1bmRlZmluZWQsIGNhbm5vdCB0cmVhdCBhcyBvYmplY3Rcbi5maXJzdF9maWVsZF90eXBlID0gdHlwZV9kZWYoLmZpZWxkKVxuXG4jIGNvbnZlciB0byBvYmplY3Rcbi5maWVsZCA9IHt9XG4uc2Vjb25kX2ZpZWxkX3R5cGUgPSB0eXBlX2RlZiguZmllbGQpXG5cbiMgbm93IGNhbiB1c2UgcGF0aCBzeW50YXhcbi5maWVsZC5zdWJmaWVsZC5zdWJzdWJmaWVsZCA9IFwidmFsdWUxXCIiLCJldmVudCI6eyJtZXNzYWdlIjoiSGVsbG8gVlJMIiwiZm9vIjoiZGVsZXRlIG1lIiwiaHR0cF9zdGF0dXMiOiIyMDAifSwiaXNfanNvbmwiOmZhbHNlLCJlcnJvciI6bnVsbH0%3D)

**Reply by pront** [reply]
Hey @alexeynl, I understand why this might confusing. I tried to collect all these ideas [here](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiZm9vID0ge31cbmZvby54ID0gXCJwcmVjaW91cyBkYXRhXCJcblxuIyBUaGUgZm9sbG93aW5nIHdpbGwgdGhyb3csIGNhbm5vdCBvdmVyd3JpdGUgYSBjaGlsZDpcbiMgZXJyb3JbRTY0Ml06IHBhcmVudCBwYXRoIHNlZ21lbnQgcmVqZWN0cyB0aGlzIG11dGF0aW9uXG4jIGZvby54LnkgPSBcIndyaXRlIG9uIHRvcCBvZiB4XCJcblxuZm9vLnggPSB7XCJhXCI6IDEgfSAjIG92ZXJ3cml0ZXMgdGhlIGZvby54IHN0cmluZ1xuZm9vLnguYiA9IDIgIyB2YWxpZCBhc3NpZ25tZW50IFxuLnJlc3VsdF8xID0gZm9vXG5cbiMgVGhlIGZvbGxvd2luZyB3aWxsIHRocm93OlxuIyBmb28ueC5iLmEgPSAzXG5cbmZvbyA9IFwidGhpcyB3aWxsIG92ZXdyaXRlIGV2ZXJ5dGhpbmdcIlxuLnJlc3VsdF8yID0gZm9vXG5cbiMgdmFsaWQgYXNzaWdubWVudFxuLmEuYiA9IDFcblxuIyB3aWxsIG92ZXJ3cml0ZSB0aGUgMVxuLnJlc3VsdF8zID0gc2V0ISguLCBbXCJhXCIsIFwiYlwiLCBcImNcIl0sIFwibmV3XCIpIiwiZXZlbnQiOnt9LCJpc19qc29ubCI6ZmFsc2UsImVycm9yIjpudWxsfQ%3D%3D). Hope this helps. You will have to instantiate an object `{}` to avoid the error.

**Reply by alexeynl** [reply]
It looks like i found the reason why it is happened from the desing point of view. Please check this [VRL Playground example](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiLm5ld2ZpZWxkID0gXCJ2YWx1ZVwiXG50MSA9IHR5cGVfZGVmKC4pXG5kZWwoLm5ld2ZpZWxkKVxuLm5ld2ZpZWxkID0gZGVsKC4pXG4uc2Vjb25kX3Jvb3RfdHlwZSA9IHR5cGVfZGVmKC4pXG4uZmlyc3Rfcm9vdF90eXBlID0gdDFcblxuIyB1bmRlZmluZWQsIGNhbm5vdCB0cmVhdCBhcyBvYmplY3Rcbi5maXJzdF9maWVsZF90eXBlID0gdHlwZV9kZWYoLmZpZWxkKVxuXG4jIGNvbnZlciB0byBvYmplY3Rcbi5maWVsZCA9IHt9XG4uc2Vjb25kX2ZpZWxkX3R5cGUgPSB0eXBlX2RlZiguZmllbGQpXG5cbiMgbm93IGNhbiB1c2UgcGF0aCBzeW50YXhcbi5maWVsZC5zdWJmaWVsZCA9IFwidmFsdWUxXCIiLCJldmVudCI6eyJtZXNzYWdlIjoiSGVsbG8gVlJMIiwiZm9vIjoiZGVsZXRlIG1lIiwiaHR0cF9zdGF0dXMiOiIyMDAifSwiaXNfanNvbmwiOmZhbHNlLCJlcnJvciI6bnVsbH0%3D), i have change your previouse example slightly.

If you look at root definition before del(.):

```
"first_root_type": {
  "object": {
	  "newfield": {
		  "bytes": true
	  }
  },
  "object_unknown_infinite": { <<---- EXISTS
	  "any": true
  }
},
```

and after

```
"second_root_type": {
"object": {
	"newfield": {
		"object": {},
		"object_unknown_infinite": { <<<IT IS HERE NOW!
			"any": true
		}
	}
} <<<---- OBJECT_UNKNOWN_INFINITE DOES NOT EXIST FOR ROOT!


```

you can see that property `object_unknown_infinite` 
```
"object_unknown_infinite": {
	"any": true
}
```
is missing for root in second case!


Is this expected behaviour???

**Reply by alexeynl** [reply]
Difenitly i found the reason and my previouse explanation is right. [This Playgound](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiLm9sZHJvb3QgPSBkZWwoLilcblxuI3RyeSB0byB1c2UgcGF0aCBzeW50YXggaW4gbmV3IHJvb3QgbGlrZSB0aGlzXG4uZmllbGQuc3ViZmllbGQuc3Vic3ViZmllbGQgPSBcIlwiXG4jd2lsbCBnZXQgZXJyb3IgcXVlcnlpbmcgYSBmaWVsZCBvZiBhIG5vbi1vYmplY3QgdHlwZSBpcyB1bnN1cHBvcnRlZFxuXG4jYnV0IG5vdCBmb3Igb2xkIHJvb3QhXG4ub2xkcm9vdC5maWVsZC5zdWJmaWVsZC5zdWJzdWJmaWVsZCA9IFwidmFsdWVcIiIsImV2ZW50Ijp7Im1lc3NhZ2UiOiJIZWxsbyBWUkwiLCJmb28iOiJkZWxldGUgbWUiLCJodHRwX3N0YXR1cyI6IjIwMCJ9LCJpc19qc29ubCI6ZmFsc2UsImVycm9yIjpudWxsfQ%3D%3D) is a new proof:
```
.oldroot = del(.)

#try to use path syntax in new root like this
.field.subfield.subsubfield = "" << --- comment this and you look at the magic)
#will get error querying a field of a non-object type is unsupported

#but no error for old root!
.oldroot.field.subfield.subsubfield = "value"

```

**Reply by jszwedko** [reply]
This is admittedly a bit subtle, but when you do `del(.)` that sets `.` to `null` and so you can no longer access fields on it without resetting it to `{}` (an object).
